File: SmokeSaver/Services/ProductSuggestionService.cs
using SmokeSaver.Models;

namespace SmokeSaver.Services;

public sealed class ProductSuggestionService
{
    private readonly IReadOnlyList<CatalogProduct> _catalogProducts;

    public ProductSuggestionService(IReadOnlyList<CatalogProduct> catalogProducts)
    {
        _catalogProducts = catalogProducts;
    }

    public IReadOnlyList<SuggestedProduct> GetSuggestions(decimal budgetUah)
    {
        if (_catalogProducts.Count == 0)
        {
            return Array.Empty<SuggestedProduct>();
        }

        var minimumPrice = _catalogProducts.Min(product => product.PriceUah);
        var effectiveBudget = Math.Max(budgetUah, minimumPrice);

        var eligibleProducts = _catalogProducts
            .Where(product => product.PriceUah <= effectiveBudget)
            .OrderBy(product => product.PriceUah)
            .ToArray();

        if (eligibleProducts.Length == 0)
        {
            eligibleProducts = _catalogProducts.Take(3).ToArray();
        }

        var selectedProducts = new List<CatalogProduct>();
        var usedCategories = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        var topTier = eligibleProducts.Where(product => product.PriceUah >= effectiveBudget * 0.72m).ToArray();
        var middleTier = eligibleProducts.Where(product => product.PriceUah >= effectiveBudget * 0.35m && product.PriceUah < effectiveBudget * 0.72m).ToArray();
        var smallTier = eligibleProducts.Where(product => product.PriceUah < effectiveBudget * 0.35m).ToArray();

        TryPickProduct(topTier, selectedProducts, usedCategories);
        TryPickProduct(middleTier, selectedProducts, usedCategories);
        TryPickProduct(smallTier, selectedProducts, usedCategories);

        while (selectedProducts.Count < 3)
        {
            if (!TryPickProduct(eligibleProducts, selectedProducts, usedCategories))
            {
                break;
            }
        }

        return selectedProducts
            .DistinctBy(product => product.Id)
            .Take(3)
            .Select(product => new SuggestedProduct
            {
                Title = product.Title,
                PriceUah = product.PriceUah,
                Category = product.Category,
                ShortDescription = product.ShortDescription
            })
            .ToArray();
    }

    private static bool TryPickProduct(
        IReadOnlyList<CatalogProduct> source,
        ICollection<CatalogProduct> selectedProducts,
        ISet<string> usedCategories)
    {
        var usedIds = selectedProducts.Select(product => product.Id).ToHashSet(StringComparer.OrdinalIgnoreCase);

        var freshCategoryCandidates = source
            .Where(product => !usedIds.Contains(product.Id) && !usedCategories.Contains(product.Category))
            .ToArray();

        var fallbackCandidates = source
            .Where(product => !usedIds.Contains(product.Id))
            .ToArray();

        var candidatePool = freshCategoryCandidates.Length > 0 ? freshCategoryCandidates : fallbackCandidates;
        if (candidatePool.Length == 0)
        {
            return false;
        }

        var pick = candidatePool[Random.Shared.Next(candidatePool.Length)];
        selectedProducts.Add(pick);
        usedCategories.Add(pick.Category);
        return true;
    }
}

File: NitKotin.Android/app/src/main/java/com/nitkotin/android/domain/ProductSuggestionService.kt
package com.nitkotin.android.domain

import com.nitkotin.android.data.model.CatalogProduct
import com.nitkotin.android.data.model.SuggestedProduct
import kotlin.math.max
import kotlin.random.Random

class ProductSuggestionService(
    private val catalogProducts: List<CatalogProduct>,
    private val random: Random = Random.Default,
) {
    fun getSuggestions(budgetUah: Double): List<SuggestedProduct> {
        if (catalogProducts.isEmpty()) {
            return emptyList()
        }

        val minimumPrice = catalogProducts.minOf { it.priceUah }
        val effectiveBudget = max(budgetUah, minimumPrice)

        var eligibleProducts = catalogProducts
            .filter { it.priceUah <= effectiveBudget }
            .sortedBy { it.priceUah }

        if (eligibleProducts.isEmpty()) {
            eligibleProducts = catalogProducts.take(3)
        }

        val selectedProducts = mutableListOf<CatalogProduct>()
        val usedCategories = linkedSetOf<String>()

        val topTier = eligibleProducts.filter { it.priceUah >= effectiveBudget * 0.72 }
        val middleTier = eligibleProducts.filter { it.priceUah >= effectiveBudget * 0.35 && it.priceUah < effectiveBudget * 0.72 }
        val smallTier = eligibleProducts.filter { it.priceUah < effectiveBudget * 0.35 }

        tryPickProduct(topTier, selectedProducts, usedCategories)
        tryPickProduct(middleTier, selectedProducts, usedCategories)
        tryPickProduct(smallTier, selectedProducts, usedCategories)

        while (selectedProducts.size < 3) {
            if (!tryPickProduct(eligibleProducts, selectedProducts, usedCategories)) {
                break
            }
        }

        return selectedProducts
            .distinctBy { it.id }
            .take(3)
            .map {
                SuggestedProduct(
                    title = it.title,
                    priceUah = it.priceUah,
                    category = it.category,
                    shortDescription = it.shortDescription,
                )
            }
    }

    private fun tryPickProduct(
        source: List<CatalogProduct>,
        selectedProducts: MutableList<CatalogProduct>,
        usedCategories: MutableSet<String>,
    ): Boolean {
        val usedIds = selectedProducts.map { it.id.lowercase() }.toSet()
        val freshCategoryCandidates = source.filter { it.id.lowercase() !in usedIds && it.category !in usedCategories }
        val fallbackCandidates = source.filter { it.id.lowercase() !in usedIds }
        val candidatePool = if (freshCategoryCandidates.isNotEmpty()) freshCategoryCandidates else fallbackCandidates
        if (candidatePool.isEmpty()) {
            return false
        }

        val pick = candidatePool[random.nextInt(candidatePool.size)]
        selectedProducts += pick
        usedCategories += pick.category
        return true
    }
}
